lib/image-editor/class-image-editor.php

```diff
@@ -0,0 +1,260 @@
+<?php
+/**
+ * Start: Include for phase 2
+ * Image Editor: Image_Editor class
+ *
+ * @package gutenberg
+ * @since 7.x ?
+ */
+
+/**
+ * Image flip/mirror modifier.
+ */
+require_once __DIR__ . '/class-image-editor-flip.php';
+
+/**
+ * Image rotate modifier.
+ */
+require_once __DIR__ . '/class-image-editor-rotate.php';
+
+/**
+ * Image crop modifier.
+ */
+require_once __DIR__ . '/class-image-editor-crop.php';
+
+/**
+ * Image editor.
+ *
+ * @since 7.x ?
+ */
+class Image_Editor {
+	const META_KEY = 'image-editor';
+
+	/**
+	 * Constructs an Image_Editor.
+	 */
+	public function __construct() {
+		$this->all_modifiers = array(
+			'Image_Editor_Crop',
+			'Image_Editor_Flip',
+			'Image_Editor_Rotate',
+		);
+	}
+
+	/**
+	 * Modifies an image.
+	 *
+	 * @param integer $media_id Media id.
+	 * @param array   $modifiers Modifier to apply to the image.
+	 * @return array|WP_Error If successful image JSON containing the media_id and url of modified image, otherwise WP_Error.
+	 */
+	public function modify_image( $media_id, $modifiers ) {
+		// Get image information.
+		$info = $this->load_image_info( $media_id );
+		if ( is_wp_error( $info ) ) {
+			return $info;
+		}
+
+		foreach ( $modifiers as $modifier ) {
+			// Update it with our modifier.
+			$info['meta'] = $modifier->apply_to_meta( $info['meta'] );
+		}
+
+		// Generate filename based on current attributes.
+		$target_file = $this->get_filename( $info['meta'] );
+
+		// Does the image already exist?
+		$existing_image = $this->get_existing_image( $info, $target_file );
```

Comment:
Instead of doing this, can't you jsut check it isn't in the list of file names in `wp_get_attachment_metadata`? Calls to `attachment_url_to_postid` should be avoided at all costs, as this is uncachable / filterable call.

Reply:
If it is a new attachment, a link between the old and new should be a created.